ArchiSteamFarm/Plugins/Core.cs

```diff
@@ -0,0 +1,290 @@
+using System.Composition.Convention;
+using System.Composition.Hosting;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using ArchiSteamFarm.Localization;
+using Newtonsoft.Json.Linq;
+using SteamKit2;
+
+namespace ArchiSteamFarm.Plugins {
+	internal static class Core {
+		[ImportMany]
+		private static ImmutableHashSet<IPlugin> ActivePlugins { get; set; }
+
+		internal static bool InitPlugins() {
+			string pluginsPath = Path.Combine(SharedInfo.HomeDirectory, SharedInfo.PluginsDirectory);
+
+			if (!Directory.Exists(pluginsPath)) {
+				ASF.ArchiLogger.LogGenericTrace(Strings.NothingFound);
+
+				return true;
+			}
+
+			ASF.ArchiLogger.LogGenericInfo(string.Format(Strings.Initializing, nameof(Plugins)));
+
+			HashSet<Assembly> assemblies = new HashSet<Assembly>();
+
+			try {
+				foreach (string assemblyPath in Directory.EnumerateFiles(pluginsPath, "*.dll")) {
+					Assembly assembly;
+
+					try {
+						assembly = Assembly.LoadFrom(assemblyPath);
+					} catch (Exception e) {
+						ASF.ArchiLogger.LogGenericException(e);
+
+						continue;
+					}
+
+					assemblies.Add(assembly);
+				}
+			} catch (Exception e) {
+				ASF.ArchiLogger.LogGenericException(e);
+
+				return false;
+			}
+
+			if (assemblies.Count == 0) {
+				ASF.ArchiLogger.LogGenericInfo(Strings.NothingFound);
+
+				return true;
+			}
+
+			ConventionBuilder conventions = new ConventionBuilder();
+			conventions.ForTypesDerivedFrom<IPlugin>().Export<IPlugin>();
+
+			ContainerConfiguration configuration = new ContainerConfiguration().WithAssemblies(assemblies, conventions);
+
+			try {
+				using (CompositionHost container = configuration.CreateContainer()) {
+					ActivePlugins = container.GetExports<IPlugin>().ToImmutableHashSet();
+				}
+			} catch (Exception e) {
+				ASF.ArchiLogger.LogGenericException(e);
+
+				return false;
+			}
+
+			HashSet<IPlugin> invalidPlugins = new HashSet<IPlugin>();
+
+			foreach (IPlugin plugin in ActivePlugins) {
+				try {
+					string pluginName = plugin.GetName();
+
+					ASF.ArchiLogger.LogGenericInfo(string.Format(Strings.PluginLoading, pluginName, plugin.GetVersion()));
+					plugin.OnLoaded();
+					ASF.ArchiLogger.LogGenericInfo(string.Format(Strings.PluginLoaded, pluginName));
+				} catch (Exception e) {
+					ASF.ArchiLogger.LogGenericException(e);
+					invalidPlugins.Add(plugin);
+				}
+			}
+
+			ImmutableHashSet<IPlugin> activePlugins = ActivePlugins.Except(invalidPlugins);
+
+			if (activePlugins.Count == 0) {
+				ActivePlugins = null;
+
+				return false;
+			}
+
+			ActivePlugins = activePlugins;
+			ASF.ArchiLogger.LogGenericInfo(Strings.PluginsWarning);
+
+			return invalidPlugins.Count == 0;
+		}
+
+		internal static void OnASFInitModules(IReadOnlyDictionary<string, JToken> additionalConfigProperties = null) {
+			if ((ActivePlugins == null) || (ActivePlugins.Count == 0)) {
+				return;
+			}
+
+			foreach (IASF plugin in ActivePlugins.OfType<IASF>()) {
```

Comment:
especially here (and below in the other functions) I'd like to see some parallelization

Reply:
problem is if there are several plugins loaded and one takes infinitely to finish the other ones would not get called at all

Reply:
A plugin that does excessive amount of async work (aka can block) is supposed to have `async void` return which will not block the calling code.

But regardless it's a good idea to start them all at the same time.

Reply:
It's not about what it is supposed to have. It's about problems that might occur when a bad coder writes bad code inside his plugin

Reply:
> when a bad coder writes bad code inside his plugin

No longer our concern.

Reply:
Of course it's not... but still

Reply:
> but still

Still you're not going to stop bad coder writing a bad code of setting all internal ASF properties to `null` via reflection and causing fatal ASF crashes during runtime.

Reply:
since when do bad coders know about [reflection](https://www.youtube.com/watch?v=h8GUCQQZS64)?